### PR TITLE
Restructure cardano testnet 

### PR DESCRIPTION
**Pull request by Jimbo4350**
# Description

Resolves: #5319

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
  - integration tests
  See [Runnings tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-testnet/src/Testnet/Filepath.hs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  :: TmpAbsolutePath
+  -> String -- ^ node name
+  -> Sprocket
+makeSprocket tmpAbsPath node
+  = Sprocket (makeTmpBaseAbsPath tmpAbsPath) (makeSocketDir tmpAbsPath </> node)
+
+-- Temporary path used at runtime
+newtype TmpAbsolutePath = TmpAbsolutePath
+  { unTmpAbsPath :: FilePath
+  } deriving (Eq, Show)
```
Comment (carbolymer) [inline cardano-testnet/src/Testnet/Filepath.hs:36]: I think having `Monoid` and `IsString` would be useful for this.

File: cardano-testnet/src/Testnet/Property/Assert.hs [hunk 2]
```diff
@@ -47,13 +50,28 @@ fileJsonGrep fp f = do
   let jsons = mapMaybe (Aeson.decode @Value) lines
   return $ L.any f jsons
 
+assertByDeadlineIOCustom
+  :: (MonadTest m, MonadIO m, HasCallStack)
+  => String -> DTC.UTCTime -> IO Bool -> m ()
+assertByDeadlineIOCustom str deadline f = GHC.withFrozenCallStack $ do
+  success <- H.evalIO f
```
Comment (carbolymer) [inline cardano-testnet/src/Testnet/Property/Assert.hs:57]: I think you could fork here and wait for the completion or kill the thread using [`timeout`](https://hackage.haskell.org/package/base-4.18.0.0/docs/System-Timeout.html#v:timeout) for example. Is there a point in waiting the `f` exceeds the deadline?

**Review by carbolymer — APPROVED** [review]

File: cardano-testnet/src/Testnet/Filepath.hs (hunk 1 above)
Comment (carbolymer) [reply cardano-testnet/src/Testnet/Filepath.hs:36]: I think `Semigroup` should be enough. For easier concatenation, for example:
```haskell
let prefix = "/tmp/tmpfile" :: TmpAbsolutePath
in prefix <> ".1234.tmp"
```

Comment (Jimbo4350) [reply cardano-testnet/src/Testnet/Filepath.hs:36]: I'll add `IsString` but I'd rather be explicit and use `</>` when concatenating `FilePath`s

**Pull request merged**